fix watcher queue initialisation on concurrency

chmouel opened the pull request:
Resolved an issue where pipelineRuns were being added to the queue instead of the running queue during the initialization of the queue manager. Introduced a new method, AddToPendingQueue, to add pipelineRuns directly to the pending queue without mixing them with the running queue.

Fixes #1790

See video below for a demo

https://github.com/user-attachments/assets/31dd535b-11e9-4b29-a4f9-f397022ad389


# Changes <!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

# Submitter Checklist

- [ ] 📝 Please ensure your commit message is clear and informative. For guidance on crafting effective commit messages, refer to the How to write a git commit message guide. We prefer the commit message to be included in the PR body itself rather than a link to an external website (ie: Jira ticket).

- [ ] ♽ Before submitting a PR, run make test lint to avoid unnecessary CI processing. For an even more efficient workflow, consider installing [pre-commit](https://pre-commit.com/) and running pre-commit install in the root of this repository.

- [ ] ✨ We use linters to maintain clean and consistent code. Please ensure you've run make lint before submitting a PR. Some linters offer a --fix mode, which can be executed with the command make fix-linters (ensure [markdownlint](https://github.com/DavidAnson/markdownlint) and [golangci-lint](https://github.com/golangci/golangci-lint) tools are installed first).

- [ ] 📖 If you're introducing a user-facing feature or changing existing behavior, please ensure it's properly documented.

- [ ] 🧪 While 100% coverage isn't a requirement, we encourage unit tests for any code changes where possible.

- [ ] 🎁 If feasible, please check if an end-to-end test can be added. See [README](https://github.com/openshift-pipelines/pipelines-as-code/blob/main/test/README.md) for more details.

- [ ] 🔎 If there's any flakiness in the CI tests, don't necessarily ignore it. It's better to address the issue before merging, or provide a valid reason to bypass it if fixing isn't possible (e.g., token rate limitations).

chmouel committed:
fix watcher queue initialisation on concurrency

Resolved an issue where pipelineRuns were being added to the queue
instead of the running queue during the initialization of the queue
manager. Introduced a new method, AddToPendingQueue, to add pipelineRuns
directly to the pending queue without mixing them with the running queue.

Fixes #1790

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

zakisk commented on pkg/sync/queue_manager.go:
```diff
@@ -114,6 +114,23 @@ func (qm *QueueManager) AddListToQueue(repo *v1alpha1.Repository, list []string)
+
+	for _, pr := range list {
+		if sema.addToPendingQueue(pr, time.Now()) {
+			qm.logger.Infof("added pipelineRun (%s) to queue for repository (%s)", pr, repoKey(repo))
```
```suggestion
			qm.logger.Infof("added pipelineRun (%s) to pending queue for repository (%s)", pr, repoKey(repo))
```

chmouel committed:
Add e2e test for pending prs processed

Add a new test that will test that when the watcher is down and a PR is
created, the PipelineRuns are kept in Pending state. When the watcher is
up again, the PipelineRuns are restarted.

Some refactoring has been done to move some functions to their own.

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

chmouel commented:
Added e2e test, which seems to work properly!
